src/webui/api/searchcontroller.cpp

```diff
@@ -203,15 +203,16 @@ void SearchController::deleteAction()
 
 void SearchController::categoriesAction()
 {
-    QStringList categories;
     const QString name = params()["pluginName"].trimmed();
 
-    categories << SearchPluginManager::categoryFullName("all");
-    for (const QString &category : asConst(SearchPluginManager::instance()->getPluginCategories(name)))
-        categories << SearchPluginManager::categoryFullName(category);
+    QJsonArray categories = getPluginCategories(SearchPluginManager::instance()->getPluginCategories(name));
+    const QJsonObject allCategory = {
+        {"id", "all"},
+        {"name", SearchPluginManager::categoryFullName("all")}
+    };
+    categories.prepend(allCategory);
```

Comment:
I think that leaves two options, neither of which seem like a win:

1. Copy the `categories` array to a new array, with `allCategory` as the first item
2. Have `getPluginCategories` accept an array to append to, which seems like a terrible interface

Reply:
>Copy the categories array to a new array, with allCategory as the first item

I still think it is slightly better than the current as it avoids shifting (actually shifting implies copying unless it is a list).
@glassez any opinions?
> Have getPluginCategories accept an array to append to, which seems like a terrible interface

Indeed, let's not consider this.

Reply:
>@glassez any opinions?

Every plugins support this metacategory ("all") so why send it? Let client append it to its list by default (it still does the same thing in some cases).

Reply:
>Every plugins support this metacategory ("all") so why send it? Let client append it to its list by default (it still does the same thing in some cases).

But WebAPI users won't know it if they only didn't consult the api doc, isn't?

Reply:
Yes. But they don't know anything without consulting the API doc.
BTW, it still does the same thing in some cases, as I said before.

Reply:
Following [POLA](https://en.wikipedia.org/wiki/Principle_of_least_astonishment), I think it makes more sense for the categories API to return all categories, including "all".  The web app does currently have a hardcoded reference to "all", but that's mostly because of a poor design choice I made. I don't think we should use that example to justify removing "all" from our API.

Reply:
>Following POLA, I think it makes more sense for the categories API to return all categories, including "all".

Well, if you put it that way...
Seems "all" isn't category. It just a placeholder. And it makes sence in single direction only, IMO.
Just for example from real life. When people ask you what you're wearing, you just list all these items of clothing, but you don't add "all" to the list. But your interlocutor, on the contrary, may ask you to change "all" your clothes.
But if that doesn't convince you, I won't push it... this issue has a low priority for me so that I spend too much effort.

Reply:
I would disagree that "all" isn't a category, mainly because specifying a category is required when performing a search. In my mind, this requirement essentially upgrades "all" to a full category.

I've pushed up a new commit that I think handles this whole situation better than before. The `/search/categories` API is removed entirely in place of calculating categories for "All plugins" and "Enabled plugins" client-side. I think this is how the web api/ui should've been handled from the start.

If reviewers are in favor of the change, I'll squash the commits.

Reply:
I also think that "all" should be considered a "full" category.

By the way, I think a better clothes analogy would be:
A: I want you to take off your clothes
B: Ok, all, just trousers, or just shirt?
A: All

Reply:
@FranciscoPombal
Sorry, my example uses characters and objects from real life, but it still has a formal logic. Yours isn't. API's shouldn't behave like this.
>I would disagree that "all" isn't a category, mainly because specifying a category is required when performing a search.

It can be drawback of current implementation.
>In my mind, this requirement essentially upgrades "all" to a full category.

If "all" is yet another category then it's supposed the plugin may not support it, isn't it?

@Chocobo1, what is your opinion? I don't want to continue this discussion if all others don't share my point.

Reply:
@Piccirello, since you don't propagate it in API, it looks good for me.

Reply:
>@Piccirello, since you don't propagate it in API, it looks good for me.

Oops... you does it.

Reply:
>@Chocobo1, what is your opinion? I don't want to continue this discussion if all others don't share my point.

I would like to see `all` is shown when invoking the webAPI. IMO webapi users should able to see every available options for `categories` field on query.

Reply:
Well, let's proceed with it.